### ISSUE TITLE
U-Net - Prediction all black

### ISSUE THREAD

**Issue by NuriaTaberner** [HenriquesLab/ZeroCostDL4Mic#21]
Hello, 

Before the last update I trained a U-Net with 48 images of my own data with their labels (tif - 8-bit, 640x640) and I got a reasonably good prediction. 
![Captura](https://user-images.githubusercontent.com/15859461/83100608-4f838700-a0eb-11ea-8d13-44c35acf040f.PNG)
I didn't save the model. 

Now, after the last update, I tried to train again U-Net, but every time I get a black prediction. Before I had tried with new training samples and also I had run it in my own computer. But eventually I returned to run exactly the same as before (or so I think) but still I do not get predictions. 

Could it be something changed? or maybe I am missing something now, that before I did right...

![Captura-14](https://user-images.githubusercontent.com/15859461/83100815-be60e000-a0eb-11ea-8e73-ce206cc17d0c.PNG)

I don't have the loss plot from before, but it had some nice evolution. However now it flattens very rapidly... 
![índex](https://user-images.githubusercontent.com/15859461/83100888-ecdebb00-a0eb-11ea-8410-e89babb21801.png)

Some outputs:
Model: "model_3"
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
==================================================================================================
input_3 (InputLayer)            (None, 640, 640, 1)  0                                            
__________________________________________________________________________________________________
conv2d_27 (Conv2D)              (None, 640, 640, 64) 640         input_3[0][0]                    
__________________________________________________________________________________________________
conv2d_28 (Conv2D)              (None, 640, 640, 64) 36928       conv2d_27[0][0]                  
__________________________________________________________________________________________________
max_pooling2d_5 (MaxPooling2D)  (None, 320, 320, 64) 0           conv2d_28[0][0]                  
__________________________________________________________________________________________________
conv2d_29 (Conv2D)              (None, 320, 320, 128 73856       max_pooling2d_5[0][0]            
__________________________________________________________________________________________________
conv2d_30 (Conv2D)              (None, 320, 320, 128 147584      conv2d_29[0][0]                  
__________________________________________________________________________________________________
max_pooling2d_6 (MaxPooling2D)  (None, 160, 160, 128 0           conv2d_30[0][0]                  
__________________________________________________________________________________________________
conv2d_31 (Conv2D)              (None, 160, 160, 256 295168      max_pooling2d_6[0][0]            
__________________________________________________________________________________________________
conv2d_32 (Conv2D)              (None, 160, 160, 256 590080      conv2d_31[0][0]                  
__________________________________________________________________________________________________
up_sampling2d_5 (UpSampling2D)  (None, 320, 320, 256 0           conv2d_32[0][0]                  
__________________________________________________________________________________________________
conv2d_33 (Conv2D)              (None, 320, 320, 128 131200      up_sampling2d_5[0][0]            
__________________________________________________________________________________________________
concatenate_5 (Concatenate)     (None, 320, 320, 256 0           conv2d_30[0][0]                  
                                                                 conv2d_33[0][0]                  
__________________________________________________________________________________________________
conv2d_34 (Conv2D)              (None, 320, 320, 128 295040      concatenate_5[0][0]              
__________________________________________________________________________________________________
up_sampling2d_6 (UpSampling2D)  (None, 640, 640, 128 0           conv2d_34[0][0]                  
__________________________________________________________________________________________________
conv2d_35 (Conv2D)              (None, 640, 640, 64) 32832       up_sampling2d_6[0][0]            
__________________________________________________________________________________________________
concatenate_6 (Concatenate)     (None, 640, 640, 128 0           conv2d_28[0][0]                  
                                                                 conv2d_35[0][0]                  
__________________________________________________________________________________________________
conv2d_36 (Conv2D)              (None, 640, 640, 64) 73792       concatenate_6[0][0]              
__________________________________________________________________________________________________
conv2d_37 (Conv2D)              (None, 640, 640, 64) 36928       conv2d_36[0][0]                  
__________________________________________________________________________________________________
conv2d_38 (Conv2D)              (None, 640, 640, 2)  1154        conv2d_37[0][0]                  
__________________________________________________________________________________________________
conv2d_39 (Conv2D)              (None, 640, 640, 1)  3           conv2d_38[0][0]                  
==================================================================================================
Total params: 1,715,205
Trainable params: 1,715,205
Non-trainable params: 0
__________________________________________________________________________________________________
None
{'lr': 0.0003000000142492354, 'beta_1': 0.8999999761581421, 'beta_2': 0.9990000128746033, 'decay': 0.0, 'epsilon': 1e-07, 'amsgrad': False}
!! WARNING: Model folder already existed and has been removed !!
---------------------------- Main training parameters ----------------------------
Number of epochs: 200
Batch size: 4
Number of training dataset: 96
Number of training steps: 24
Number of validation steps: 3
---------------------------- ------------------------ ----------------------------
I am using the augmentation options without vertical flip.

Thank you in advance, 

Nuria

**Comment by lucpaul** [comment]
Hi,
yes, we made quite a few changes to the Unet notebook. The main one from what I can see in your example is that in the new notebook we are converting the raw images to segmentation masks instead of saving just the raw image. So the output that you will see is a binary mask now.

What you could try is to test your quality control dataset in the prediction cell (6). The output of that cell should show you the predicted raw image beside the predicted mask that is made from the image. Can you have a look if this image is completely black as well? If so, there might be something happening in the actual Unet, otherwise, it might simply have to do with the thresholding that is used to create the mask.
You should also be able to play around with the thresholding a little bit. Does that help?

Another thing you can try is to reduce the augmentation and see if that changes how well the network performs. We did change the way we feed the data to the model since you last used the notebook, and have noticed that the augmentation can make quite a difference. Try to reduce some of the augmentations and see what happens.

**Comment by NuriaTaberner** [comment]
Dear lucPaul, 

Thank you for the reply. 

- It doesn't seem to be a matter of the threshold since the raw prediction is all zeroes. 
- I tried 50 epochs with smaller augmentation options. 0.01 shift, 30 rotation, shear 0.02. I got the same result. I am now running for 200 epochs with in between values: 0.05 shift, 30 rotation, shear 0.05.

**Comment by Romain-Laine** [comment]
Hi NuriaTaberner,
I think I know what might be happening there. It looks like your masks only look at very small parts of the image and therefore, the model will tend to predict that nothing should be thresholded because it's a fairly good solution in the end if 99% of your image should be background. It's a typical case of class imbalance where there's a lot more of class 0 than class 1. 
As a matter of fact, in the new version (due to be released next week), we have implemented a weighted binary cross-entropy loss function to address that !

Also, the new version will implement the predictions differently, which should also improve performance there. 
Could you please give us more info about which release you were using before and which release you are using now? This way we can track when the issues appeared.

Thanks !

Romain

**Comment by NuriaTaberner** [comment]
Dear Romain, 

Thank you very much for the response. 
The version I had used before was the one that was on github just right before than 13 days ago. So right before the last update published. 
I look forward to the new version, however, would it be possible to have access to the previous one also?

Thank you, 

Nuria

**Comment by Romain-Laine** [comment]
Hi Nuria, I think that the weighted binary cross-entropy will still be helpful for you in the next release so you should try. But I am glad to hear that fine-tuning the LR could help you converge better. As for the versions, all previously provided versions are available via the "Release" page of this GitHub repo. We do not upload notebooks outside of releases so the previous version you'd been using must be one of the releases. Go and have a look ! 
If all is good, please let me know and I'll close this issue!
Cheers,

Romain

**Comment by NuriaTaberner** [comment]
Dear Romain, 

I am looking forward to the new release with weighted binary cross-entropy. 

I'm also getting good results now and nice loss curves. 

Therefore, you can close this threat. 

Thank you, 

Nuria
